### DISCUSSION TITLE
PostCSS plugin tailwindcss requires PostCSS 8.

### DISCUSSION THREAD

**Discussion by imohuan** [tailwindlabs/tailwindcss#2854] [Help] [answered]
## Either the corresponding version installed or the latest version will still report an error
``` javascript
Module build failed (from ./node_modules/postcss-loader/src/index.js):
Error: PostCSS plugin tailwindcss requires PostCSS 8.
Migration guide for end-users:
https://github.com/postcss/postcss/wiki/PostCSS-8-for-end-users
```

``` json
  "dependencies": {
    "autoprefixer": "^10.0.2",
    "postcss": "^8.1.0",
    "tailwindcss": "^2.0.1",
    "vue": "^3.0.0"
  },
```

**Comment by flakerimi** [comment]
when I did with yarn I received

```
error tailwindcss@2.0.1: The engine "node" is incompatible with this module. Expected version ">=12.13.0".
error Found incompatible module
```
Then some errors disapear.

You need to have

    "postcss": "^8.1.0",
    "postcss-loader": "^4.1.0",
    "tailwindcss": "^2.0.1",

**Reply by dario13** [reply]
thanks! solved for me adding postcss-loader

**Comment by jenstornell** [comment, marked as answer]
I had the same thing with Vue 3 today. Solve it with the below.

https://tailwindcss.com/docs/installation#post-css-7-compatibility-build

```
npm uninstall tailwindcss postcss autoprefixer
npm install tailwindcss@npm:@tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9
```

**Reply by djasnowski** [reply]
This worked for me using Parcel `1.12.4`.

**Reply by aminesmkhani** [reply]
tnx good job its working vue3

**Reply by Androz2091** [reply]
Is it still the recommended solution?

**Reply by flakerimi** [reply]
@Androz2091 
Add this to package Json 
```
"postcss": "^8.1.0",
"postcss-loader": "^4.1.0",
"tailwindcss": "^2.0.1",
```

Without `"postcss-loader"` shows that error.

**Reply by Androz2091** [reply]
I'm still getting:
```
[vite] SFC style compilation error: 
/Users/simon/Documents/GitHub/BetterCommands/src/components/CommandGroup.vue
PostCSS plugin tailwindcss requires PostCSS 8.
Migration guide for end-users:
https://github.com/postcss/postcss/wiki/PostCSS-8-for-end-users
```
Anyway I can stay on postcss7 but I wanted to know if we could migrate to v8... it's better I guess 🤔

**Reply by vivioli** [reply]
how do i fix it? 

    % npm install tailwindcss@npm:@tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9
    npm ERR! code ERESOLVE
    npm ERR! Cannot read property 'length' of undefined

**Reply by PhilipAnd** [reply]
When running the following code:
`npm install tailwindcss@npm:@tailwindcss/postcss7-compat @tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9`

I get the error:
`no matches found: postcss@^7`

**Reply by ming-gao** [reply]
Thank you!

**Reply by zhaoqize** [reply]
thanks

**Reply by hognevevle** [reply]
Isn't Vue 3.0.6 supposed to support PostCSS 8? Or are there still other parts of Vue that still doesn't support it?

**Reply by QuadradS** [reply]
Thanks man!

**Reply by KlimDos** [reply]
God bless you boss

**Reply by sunny-mittal** [reply]
@PhilipAnd You must be using zsh or another shell that is expanding special characters. Prefix the command with `noglob` and you'll be golden.

**Reply by TreehouseFalcon** [reply]
Still works, thanks!

**Reply by Stevemoretz** [reply]
Didn't work for me.

**Reply by gabrielAlberte** [reply]
thanks dude

**Reply by xmetaki** [reply]
works on vue2

**Reply by ma7ma7** [reply]
It solve my problem to for react v 17

**Reply by neverkas** [reply]
It doesn't work at all, I've another error when I did it

`PostCSS plugin postcss-discard-comments requires PostCSS 8.`

**Reply by aspearman** [reply]
This worked for me. I did it with Yarn.

**Reply by ayodeji334** [reply]
Thanks. This worked for me too.

**Reply by mmeyer77** [reply]
thanks. this solved my error with gatsby build

**Reply by snflo16** [reply]
thanks. it also worked for me vue3

**Reply by NestInSpace** [reply]
Thank You... It worked with **Yarn** . If your are using React,  try this ------  `yarn add -D tailwindcss@npm:@tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9`

**Reply by sifferhans** [reply]
Is there a solution for this with Tailwind 3.0 too?

**Reply by mymakarim** [reply]
well this solution does not upgrade tailwind to v3 :(
So print still does not work

**Reply by JamshedAlamQaderi** [reply]
so much thanks. this help me get rid of that postcss 8 requires error.

**Reply by Hugo-osorio** [reply]
Here is how I solved this issue with Tailwind 3.0:
- I upgraded webpack version Storybook uses as shown [here](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#webpack-5) 
`npm install @storybook/builder-webpack5 @storybook/manager-webpack5 --save-dev`
Then edit .storybook/main.js:
`module.exports = {
  core: {
    builder: 'webpack5',
  },
};`
- I upgraded the postcss-loader to version ^6:
`npm install postcss-loader@^6`
- I modified how Storybook uses the API of postcss-loader in `.storybook/main.js`:
modify this: `plugins: [*your-plugins*]` with this: `postcssOptions: { plugins: [*your-plugins*] }` 
Hope this helps anyone struggling with this :)

**Reply by saadazghour** [reply]
Thanks. This worked for me !!

**Reply by adhekeeloho** [reply]
This fixed my problem. Thank you so so much

**Reply by shahid92-test** [reply]
> I had the same thing with Vue 3 today. Solve it with the below.
> 
> https://tailwindcss.com/docs/installation#post-css-7-compatibility-build
> 
> ```
> npm uninstall tailwindcss postcss autoprefixer
> npm install tailwindcss@npm:@tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9
> ```

I don't want to downgrade tailwindCSS to version 2.

**Reply by nikavolyagoalie** [reply]
Спасибо! Помогло

**Reply by paulorobertoalmeida** [reply]
Thanks! Now everything is working smoothly 👍🏾

**Reply by sercanbayrambey** [reply]
saved the day!

**Reply by bhukyavenkatamahesh** [reply]
Thanks a lot ! It worked for me.

**Comment by Debdut** [comment]
I got the flooring error when I ran

```
npm install tailwindcss@npm:@tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9
```

```sh
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! Found: postcss@7.0.35
npm ERR! node_modules/postcss
npm ERR!   dev postcss@"^7" from the root project
npm ERR!   postcss@"^7.0.32" from autoprefixer@9.8.6
npm ERR!   node_modules/autoprefixer
npm ERR!     dev autoprefixer@"^9" from the root project
npm ERR!     autoprefixer@"^9.6.0" from preact-cli@3.0.5
npm ERR!     node_modules/preact-cli
npm ERR!       dev preact-cli@"^3.0.0" from the root project
npm ERR!   40 more (critters-webpack-plugin, css-declaration-sorter, ...)
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer postcss@"^8.1.13" from postcss-nested@5.0.5
npm ERR! node_modules/postcss-nested
npm ERR!   postcss-nested@"5.0.5" from tailwindcss@2.1.1
npm ERR!   node_modules/tailwindcss
npm ERR!     peer tailwindcss@"2.x.x" from preact-cli-tailwind@2.0.1
npm ERR!     node_modules/preact-cli-tailwind
npm ERR!       dev preact-cli-tailwind@"^2.0.1" from the root project
npm ERR!     1 more (the root project)
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 

```

**Reply by gianValentin** [reply]
Add --force

npm install tailwindcss@npm:@tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9 --force

**Reply by mgav** [reply]
Hmm. Any possible negative unintended consequences from adding "--force"?

**Comment by mgav** [comment]
This post  https://github.com/tailwindlabs/tailwindcss/discussions/3575 suggests the possible need to escape `^` by adding `\` before each instance, so the revised code would be:

`npm install -D tailwindcss@npm:@tailwindcss/postcss7-compat @tailwindcss/postcss7-compat postcss@\^7 autoprefixer@\^9`

**Reply by mgav** [reply]
plus `--force` at the end, if you are so inclined

**Comment by mgav** [comment]
Well, I decided to try the proposed solution and tried `npm uninstall tailwindcss postcss autoprefixer` per tailwind docs,

then `npm install -D tailwindcss@npm:@tailwindcss/postcss7-compat @tailwindcss/postcss7-compat postcss@\^7 autoprefixer@\^9 --force`

which resulted in this error:

```
ERROR  Failed to compile with 1 error                                                                  11:01:59 AM

 error  in ./src/App.vue?vue&type=style&index=0&id=7ba5bd90&lang=css

Syntax Error: Error: PostCSS plugin tailwindcss requires PostCSS 8.
Migration guide for end-users:
https://github.com/postcss/postcss/wiki/PostCSS-8-for-end-users


 @ ./node_modules/vue-style-loader??ref--6-oneOf-1-0!./node_modules/css-loader/dist/cjs.js??ref--6-oneOf-1-1!./node_modules/vue-loader-v16/dist/stylePostLoader.js!./node_modules/postcss-loader/src??ref--6-oneOf-1-2!./node_modules/cache-loader/dist/cjs.js??ref--0-0!./node_modules/vue-loader-v16/dist??ref--0-1!./src/App.vue?vue&type=style&index=0&id=7ba5bd90&lang=css 4:14-360 15:3-20:5 16:22-368
 @ ./src/App.vue?vue&type=style&index=0&id=7ba5bd90&lang=css
 @ ./src/App.vue
 @ ./src/main.js
 @ multi (webpack)-dev-server/client?http://192.168.1.244:8080&sockPath=/sockjs-node (webpack)/hot/dev-server.js ./src/main.js
```

BTW, I'm using Vue 3.

From my package.json, after running the above:

```  "devDependencies": {
    "@tailwindcss/postcss7-compat": "^2.2.10",
    "@vue/cli-plugin-babel": "~4.5.0",
    "@vue/cli-plugin-router": "~4.5.0",
    "@vue/cli-service": "~4.5.0",
    "@vue/compiler-sfc": "^3.0.0",
    "autoprefixer": "^9.8.6",
    "postcss": "^7.0.36",
    "postcss-cli": "^8.3.1",
    "tailwindcss": "npm:@tailwindcss/postcss7-compat@^2.2.10"
  }```

**Comment by sagunji** [comment]
I was facing the same problem.
I followed the configuration in the repo: https://github.com/wagerfield/storybook-tailwind
I did have to make changes for typescript.
```
// postcss.config.js

module.exports = {
  plugins: {
    tailwindcss: {},
    autoprefixer: {},
  },
};

```

```
// .storybook/main.js

module.exports = {
  stories: ['../src/**/*.stories.mdx', '../src/**/*.stories.@(js|jsx|ts|tsx)'],
  addons: [
    '@storybook/addon-essentials',
    {
      name: '@storybook/addon-postcss',
      options: {
        postcssLoaderOptions: {
          implementation: require('postcss'),
        },
      },
    },
  ],
};


```

**Reply by bas00445** [reply]
This config solved my problem. Thank you

**Comment by hmbilal** [comment]
# Without Downgrading and Using Vue, Tailwind 3

 1. Create `tailwind.config.js`
 2. Create `postcss.config.js`
 3. Add following in `postcss.config.js`

```
module.exports = {
  plugins: {
    tailwindcss: {
      config: './tailwind.config.js'
    }
  }
}
```

The above configuration works fine with simple npm/yarn, latest postcss, autoprefixer, vue 3 and tailwind 3 for me.

**Reply by BryanHunt** [reply]
I was able to get it to work with this configuration, but hot load doesn't work correctly.  You have to restart the server.

**Reply by BryanHunt** [reply]
Switching to vite fixed everything.

**Reply by leoVolk** [reply]
Would you mind sharing a version of your project with a working config? Switching to vite just ended up crashing the build process. 

EDIT: Forget it, got it working upgrading the vue-cli. :)

**Reply by mahendranviji333** [reply]
> The above configuration works fine with simple npm/yarn, latest postcss, autoprefixer, vue 3 and tailwind 3 for me.

thank you this working fine this only way without downgrade tailwindcss its working good

**Comment by leoVolk** [comment]
When working on a project using the vue cli, I fixed it by using the beta version of the vue cli and following the upgrade version [here](https://github.com/forsartis/vue-cli-plugin-tailwind/releases/tag/v3.0.0-beta.0)

**Comment by sugionosihuahua** [comment]
in case if someone still get this error when do command `npm run build` you can use vite build instead of  vue-cli build.
btw my project using vue 3.2.25 and vite 5.2.11 while running in node 16.
the steps :
run command `npm install vite`
then create new file in the root with name `vite.config.mjs` and put this code inside

```js
import vue from '@vitejs/plugin-vue'

export default {
  plugins: [vue()],
  base: './',
}
```
**NOTE: don't forget to change the base to the folder that contains your index.html file**

**_or you can create a new vite project and copy necessary files to your current project._**

and then open package.json (you can backup it just in case)
look for the scripts line, replace the current build command (vue-cli) to the vite.
from this :
` "build": "vue-cli-service build" `
to this :
` "build": "vite build" `

and if you want to run `npm run dev` don't forget to add this in the script line
`"dev": "vite"`
